client/Collision.coffee

```diff
@@ -0,0 +1,121 @@
+import {Aabb} from './Dbvt'
+
+intersectsSpecific = 
+    pen: (aabb, obj) ->
+        # Lines first
+        for i in [0...(obj.pts.length - 1)]
+            pt0 = obj.pts[i]
+            pt1 = obj.pts[i + 1]
+            if pt0.x == pt1.x && pt0.y == pt1.y
+                continue
+
+            # Transform everything so the line is centered at the origin
+            center = {x: (pt1.x + pt0.x) / 2, y: (pt1.y + pt0.y) / 2}
+            pt = {x: pt0.x - center.x, y: pt0.y - center.y}
+            testPt = aabb.center()
+            testPt = {x: testPt.x - center.x, y: testPt.y - center.y}
+
+            # Fatten line into rectangle consisting of points pt0 -> pt1 -> -pt0 -> -pt1 in counterclockwise order
+            mag = Math.sqrt (pt.x * pt.x + pt.y * pt.y)
+            perp = {x: pt.y / mag * obj.width / 2, y: -pt.x / mag * obj.width / 2}
+            pt0 = {x: pt.x - perp.x, y: pt.y - perp.y}
+            pt1 = {x: pt.x + perp.x, y: pt.y + perp.y}
+
+            # Cycle points until pt0->pt1 goes in the +x and +y direction
+            while pt0.x > pt1.x || pt0.y > pt1.y
+                [pt0, pt1] = [pt1, {x: -pt0.x, y: -pt0.y}]
+
+            # For want of a vector library
+            pt2 = {x: -pt0.x, y: -pt0.y}
+            pt3 = {x: -pt1.x, y: -pt1.y}
+
+            # Test point against Minkowski sum of the rectangle and the aabb
+            # It's convex, so do intersection of half-planes
+            inHalfPlanes = true
+            for [p0, p1, fx, fy] in [
+                [pt0, pt1, 1, -1],
+                [pt1, pt2, -1, -1],
+                [pt2, pt3, -1, 1],
+                [pt3, pt0, 1, 1],
+            ]
+                # Test diagonal half-plane extended from rectangle edge
+                p2 = {x: testPt.x + fx * aabb.width() / 2, y: testPt.y + fy * aabb.height() / 2}
+                # Counterclockwise check
+                if (p1.x - p0.x) * (p2.y - p0.y) - (p1.y - p0.y) * (p2.x - p0.x) > 0
+                    inHalfPlanes = false
+                    break
+
+            # Test horizontal and vertical half-planes that come from the aabb
+            # when factored into the Minkowski sum
+            if inHalfPlanes && \
+                testPt.x + aabb.width() / 2 >= pt0.x && \
+                testPt.y - aabb.height() / 2 <= pt1.y && \
+                testPt.x - aabb.width() / 2 <= -pt0.x && \
+                testPt.y + aabb.height() / 2 >= -pt1.y
+                return true
```

Comment:
This definitely feels like a lot of code done on every segment of a pen stroke.  I wonder how long it actually takes?  And whether it can be simplified?